File: 6.Sorted Challenges  310p/6.数据结构  76/链表/25. Reverse Nodes in k-Group.ts
namespace Leetcode25{
    class ListNode{
        val:number 
        next:ListNode | null
        constructor(val:number=0,next:ListNode | null ){
            this.val =val
            this.next = next 
        }
    }
    type T =ListNode |null
    const reverse=(a:T,b:ListNode |null):T=>{
        let pre=null
        let cur=a
        let rest=a 
        while(cur && cur!==b){
            rest=cur.next 
            cur.next=pre
            pre=cur
            cur=rest
        }
        return pre 
    }

    function reverseKGroup(head:T, k:number):T{
        if(!head){ return head}
        let a=head
        let b:T=head;
        for(let _ of new Array(k)){
            if(!b){ return head }
            b=b.next
        }
        let new_head=reverse(a,b)
        a.next=reverseKGroup(b,k)
        return new_head
    }
}

File: legacy/6.Sorted Challenges  310p/6.数据结构  76/链表/25. Reverse Nodes in k-Group.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverse(self, a: ListNode, b: ListNode) -> ListNode:
        # reverse [a,b)
        pre, cur, rest = None, a, a
        while cur != b:
            rest = cur.next
            cur.next = pre
            pre = cur
            cur = rest
        return pre

    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return
        a, b = head, head
        for _ in range(k):
            if not b:
                return head
            b = b.next
        new_head = self.reverse(a, b)
        a.next = self.reverseKGroup(b, k)
        return new_head
